velox/functions/sparksql/aggregates/FirstLastAggregate.cpp

```diff
@@ -0,0 +1,353 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <fmt/format.h>
+#include <string>
+
+#include "velox/expression/FunctionSignature.h"
+#include "velox/functions/lib/SimpleNumericAggregate.h"
+#include "velox/functions/prestosql/aggregates/SingleValueAccumulator.h"
+
+namespace facebook::velox::functions::sparksql::aggregate {
+
+namespace {
+
+using namespace facebook::velox::aggregate;
+
+/// FirstLastAggregate returns the first or last value of |expr| for a group of
+/// rows. If |ignoreNull| is true, returns only non-null values.
+///
+/// The function is non-deterministic because its results depends on the order
+/// of the rows which may be non-deterministic after a shuffle.  This can be
+/// made deterministic by providing explicit ordering by adding order by or sort
+/// by in query.
+template <bool numeric, typename TDataType>
+class FirstLastAggregateBase
+    : public SimpleNumericAggregate<TDataType, TDataType, TDataType> {
+  using BaseAggregate = SimpleNumericAggregate<TDataType, TDataType, TDataType>;
+
+ public:
+  explicit FirstLastAggregateBase(TypePtr resultType)
+      : BaseAggregate(resultType) {}
+
+  int32_t accumulatorFixedWidthSize() const override {
+    if (numeric) {
+      return sizeof(TDataType);
+    } else {
+      return sizeof(SingleValueAccumulator);
+    }
+  }
+
+  void initializeNewGroups(
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    exec::Aggregate::setAllNulls(groups, indices);
+    if (!numeric) {
+      for (auto i : indices) {
+        new (groups[i] + exec::Aggregate::offset_) SingleValueAccumulator();
+      }
+    }
+  }
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    if (numeric) {
+      BaseAggregate::doExtractValues(
+          groups, numGroups, result, [&](char* group) {
+            return *exec::Aggregate::value<TDataType>(group);
+          });
+    } else {
+      VELOX_CHECK(result);
+      (*result)->resize(numGroups);
+
+      auto* rawNulls = exec::Aggregate::getRawNulls(result->get());
+
+      for (auto i = 0; i < numGroups; ++i) {
+        char* group = groups[i];
+        if (exec::Aggregate::isNull(group)) {
+          (*result)->setNull(i, true);
+        } else {
+          exec::Aggregate::clearNull(rawNulls, i);
+          auto accumulator =
+              exec::Aggregate::value<SingleValueAccumulator>(group);
+          accumulator->read(*result, i);
+        }
+      }
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    extractValues(groups, numGroups, result);
+  }
+
+ protected:
+  void updateValue(vector_size_t i, char* group, DecodedVector& decoded) {
```

Comment:
I expect this logic to be different for 'first' and 'last' functions.

'first' function would store the value only if currently there is no value.

'last' function would store the value unconditionally.

Reply:
Is any way to check whether `group` is set or not?

Loop all rows and simply update group value may better than explicitly compare and set?